test/gtest/ternary_tensor_ops.cpp

```diff
@@ -0,0 +1,452 @@
+                                const int64_t dstOffset,
+                                const int64_t src1Offset,
+                                const int64_t src2Offset)
+    {
+        const auto& dstStrides  = dstSubDesc.GetStrides();
+        const auto& src1Strides = src1SubDesc.GetStrides();
+        const auto& src2Strides = src2SubDesc.GetStrides();
+
+        const auto& src1Lens = src1SubDesc.GetLengths();
+        const auto& src2Lens = src2SubDesc.GetLengths();
+
+        auto operate_over_subtensor_impl =
+            [&, dataOp, max_dim = src1Lens.size() - 1](auto&& self,
+                                                       const size_t current_dim,
+                                                       const int64_t dstOff,
+                                                       const int64_t src1Off,
+                                                       const int64_t src2Off) -> void {
+            const auto dstStride  = dstStrides[current_dim];
+            const auto src1Stride = src1Strides[current_dim];
+            const auto src2Stride = src2Strides[current_dim];
+            const bool squashed   = src1Lens[current_dim] != src2Lens[current_dim];
+
+            int64_t dstIdx  = dstOff;
+            int64_t src1Idx = src1Off;
+            int64_t src2Idx = src2Off;
+
+            for(size_t i = 0; i < src1Lens[current_dim]; ++i)
+            {
+                if(current_dim < max_dim)
+                {
+                    self(self, current_dim + 1, dstIdx, src1Idx, src2Idx);
+                }
+                else
+                {
+                    dataOp(
+                        dstSuperTensor[dstIdx], src1SuperTensor[src1Idx], src2SuperTensor[src2Idx]);
+                }
+                dstIdx += dstStride;
+                src1Idx += src1Stride;
+                src2Idx += squashed ? 0 : src2Stride;
+            }
+        };
+        operate_over_subtensor_impl(
+            operate_over_subtensor_impl, 0, dstOffset, src1Offset, src2Offset);
+    }
+
+    void CompareResults(const tensor<T>& tensorGPU, const tensor<T>& tensorCPU)
+    {
+        const TestCase& testCase = GetParam();
+
+        double tolerance = 1;
+
+        if(std::is_same_v<T, half_float::half>)
+        {
+            // taken from original c-test
+            tolerance = 80;
+        }
+
+        double threshold = std::numeric_limits<T>::epsilon() * tolerance;
+        double error     = miopen::rms_range(tensorCPU.data, tensorGPU.data);
+
+        ASSERT_LE(error, threshold)
+            << "TensorOp: " << testCase.operation << std::endl
+            << "A tensor: " << tensorA.desc.ToString() << std::endl
+            << "B tensor: " << tensorB.desc.ToString() << std::endl
+            << "IsPacked: " << testCase.packed << std::endl
+            << "Offsets: " << testCase.offsets[0] << "," << testCase.offsets[1] << ","
+            << testCase.offsets[2] << std::endl;
```

Comment:
Usually failed test prints its testcase automatically.

Reply:
It prints, but it is not readable becauseits  in a binary format.

Reply:
It can have an `operator<<`.